### PR TITLE
IEP-371: Heap Tracing

### PR DESCRIPTION
**Pull request by alirana01**
Review for the Heap Tracing functionality

**Note: Please do not merge this PR even after approving the changes**

### CODE REVIEW

File: bundles/com.espressif.idf.core/src/com/espressif/idf/core/util/FileUtil.java [hunk 1]
```diff
@@ -302,4 +313,94 @@ public static void pipe(InputStream input, OutputStream output) throws IOExcepti
...
+			inputStream.close();
+			scanner.close();
+		}
+		catch (Exception e)
+		{
+			Logger.log(e);
+			return null;
+		}
+
+		return fileContents.toString();
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.core/src/com/espressif/idf/core/util/FileUtil.java:350]: We can replace this code with call readFile(String absoluteFilePath) as you have already implemented this

File: bundles/com.espressif.idf.debug.gdbjtag.openocd/META-INF/MANIFEST.MF [hunk 2]
```diff
@@ -26,7 +26,9 @@ Require-Bundle: ilg.gnumcueclipse.debug.gdbjtag,
  org.eclipse.cdt.build.gcc.core,
  com.espressif.idf.launch.serial.core;bundle-version="1.0.0",
  com.espressif.idf.ui;bundle-version="1.0.0",
- ilg.gnumcueclipse.core
+ ilg.gnumcueclipse.core,
+ org.eclipse.cdt.debug.ui;bundle-version="8.4.200",
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.debug.gdbjtag.openocd/META-INF/MANIFEST.MF:30]: Could you remove the version number for this? Eclipse CDT 2020-09 comes with 8.4.100 and is supported by IEP

File: bundles/com.espressif.idf.debug.gdbjtag.openocd/META-INF/MANIFEST.MF [hunk 3]
```diff
@@ -26,7 +26,9 @@ Require-Bundle: ilg.gnumcueclipse.debug.gdbjtag,
...
+ org.eclipse.swt
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.debug.gdbjtag.openocd/META-INF/MANIFEST.MF:31]: We can remove org.eclipse.swt dependency as this was exported by parents' plugins. For example org.eclipse.ui

File: bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/Configuration.java [hunk 4]
```diff
@@ -179,6 +179,20 @@ public static String getGdbClientCommand(ILaunchConfiguration configuration, Str
...
+//			{
+//				IProject project = ResourcesPlugin.getWorkspace().getRoot().getProject(configuration.getAttribute(ICDTLaunchConfigurationConstants.ATTR_PROJECT_NAME, (String)null));
+//				IFile filePath = project.getFile(new Path("build/gdbinit"));
+//				lst.add(filePath.getLocation().toString());
+//			}
+//		}
+//		catch (CoreException e1)
+//		{
+//			Logger.log(e1);
+//		}
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/Configuration.java:195]: Could you remove the commented code

File: bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/heaptracing/HeapTraceComposite.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+/**
+ * Heap tracing break point composite to draw controls
+ * 
+ * @author Ali Azam Rana
+ *
+ */
+public class HeapTraceComposite extends Composite
+{
+	public static final String COMPOSITE_NAME = "start_heap_trace_composite";
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/heaptracing/HeapTraceComposite.java:31]: Add NLS

File: bundles/com.espressif.idf.ui/META-INF/MANIFEST.MF [hunk 6]
```diff
@@ -30,12 +30,22 @@ Require-Bundle: org.eclipse.core.runtime,
...
  org.eclipse.jgit,
  org.eclipse.swtchart,
- org.eclipse.cdt.native.serial
+ com.google.gson,
+ org.eclipse.nebula.widgets.xviewer;bundle-version="1.1.0",
+ org.eclipse.nebula.widgets.xviewer.core;bundle-version="1.0.0",
+ org.eclipse.nebula.widgets.opal.dialog;bundle-version="1.0.0",
+ org.eclipse.nebula.widgets.opal.dialog.snippets;bundle-version="1.0.0",
+ org.eclipse.nebula.widgets.opal.logindialog;bundle-version="1.0.0",
+ org.eclipse.nebula.widgets.opal.logindialog.snippets;bundle-version="1.0.0",
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/META-INF/MANIFEST.MF:39]: Could you check if we need `snippets` plugins and remove them if not needed

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/HeapDumpAnalysisHandler.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+		activateIDFConsoleView();
+		List<String> commands = new ArrayList<String>();
+		commands.add(IDFUtil.getIDFPythonEnvPath());
+		commands.add(IDFUtil.getIDFSysviewTraceScriptFile().getAbsolutePath());
+		commands.add("-j");
+		commands.add("-b");
+		commands.add(elfSymbolsFile.getRawLocation().toOSString());
+		commands.add("file://".concat(dumpFile.getRawLocation().toString()));
+		console.println("Commands Prepared");
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/HeapDumpAnalysisHandler.java:59]: Add NLS/externalize whever required

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/HeapDumpAnalysisHandler.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+import com.espressif.idf.ui.tracing.HeapTracingAnalysisEditor;
+import com.espressif.idf.ui.update.AbstractToolsHandler;
+
+/**
+ * Handler class for handling the context menu action
+ * 
+ * @author Ali Azam Rana
+ *
+ */
+public class HeapDumpAnalysisHandler extends AbstractToolsHandler
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/HeapDumpAnalysisHandler.java:40]: Hmm..do you see the need to extend AbstractToolsHandler instead AbstractHandler? If it's only for console purposes I would suggest using the new IDFConsole().getConsoleStream()

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/HeapDumpAnalysisHandler.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+{
+
+	@Override
+	public Object execute(ExecutionEvent event) throws ExecutionException
+	{
+		// get the selected dumpFile
+		IResource dumpFile = EclipseHandler.getSelectedResource((IEvaluationContext) event.getApplicationContext());
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/HeapDumpAnalysisHandler.java:47]: The application context may be null.

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/HeapDumpAnalysisHandler.java [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+		commands.add(IDFUtil.getIDFSysviewTraceScriptFile().getAbsolutePath());
+		commands.add("-j");
+		commands.add("-b");
+		commands.add(elfSymbolsFile.getRawLocation().toOSString());
+		commands.add("file://".concat(dumpFile.getRawLocation().toString()));
+		console.println("Commands Prepared");
+		for (String command : commands)
+		{
+			console.println(command);
+		}
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/HeapDumpAnalysisHandler.java:63]: For better readability can we display it in a single line?

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/ITracingConstants.java [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+	String EVENTS_KEY = "events";
+
+	String HEAP_KEY = "heap";
+
+	String HEAP_ALLOC_EVENT_KEY = "alloc";
+
+	String HEAP_FREE_EVENT_KEY = "free";
+
+	String STREAMS_KEY = "streams";
+}
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/ITracingConstants.java:24]: Add NLS

**Review by kolipakakondal — COMMENTED** [review]
LGTM

File: bundles/com.espressif.idf.core/src/com/espressif/idf/core/util/FileUtil.java (hunk 1 above)
Comment (alirana01) [reply bundles/com.espressif.idf.core/src/com/espressif/idf/core/util/FileUtil.java]: Done

File: bundles/com.espressif.idf.debug.gdbjtag.openocd/META-INF/MANIFEST.MF (hunk 2 above)
Comment (alirana01) [reply bundles/com.espressif.idf.debug.gdbjtag.openocd/META-INF/MANIFEST.MF]: Done

File: bundles/com.espressif.idf.debug.gdbjtag.openocd/META-INF/MANIFEST.MF (hunk 3 above)
Comment (alirana01) [reply bundles/com.espressif.idf.debug.gdbjtag.openocd/META-INF/MANIFEST.MF]: Done

File: bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/Configuration.java (hunk 4 above)
Comment (alirana01) [reply bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/Configuration.java]: Done

File: bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/heaptracing/HeapTraceComposite.java (hunk 5 above)
Comment (alirana01) [reply bundles/com.espressif.idf.debug.gdbjtag.openocd/src/com/espressif/idf/debug/gdbjtag/openocd/heaptracing/HeapTraceComposite.java]: Done

File: bundles/com.espressif.idf.ui/META-INF/MANIFEST.MF (hunk 6 above)
Comment (alirana01) [reply bundles/com.espressif.idf.ui/META-INF/MANIFEST.MF]: Done

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/HeapDumpAnalysisHandler.java (hunk 7 above)
Comment (alirana01) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/HeapDumpAnalysisHandler.java]: Done

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/HeapDumpAnalysisHandler.java (hunk 8 above)
Comment (alirana01) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/HeapDumpAnalysisHandler.java]: Done

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/HeapDumpAnalysisHandler.java (hunk 9 above)
Comment (alirana01) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/HeapDumpAnalysisHandler.java:57]: Done

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/HeapDumpAnalysisHandler.java (hunk 10 above)
Comment (alirana01) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/HeapDumpAnalysisHandler.java]: Done

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/ITracingConstants.java (hunk 11 above)
Comment (alirana01) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/tracing/ITracingConstants.java]: Done

**Review by kolipakakondal — APPROVED** [review]
LGTM. Please also add documentation in the separate readme file under the docs folder and link it to the main page

**Pull request merged**